src/schemaVisitor.ts

```diff
@@ -0,0 +1,464 @@
+  public schema: GraphQLSchema;
+
+  // Determine if this SchemaVisitor (sub)class implements a particular
+  // visitor method.
+  public static implementsVisitorMethod(methodName: string) {
+    if (! methodName.startsWith('visit')) {
+      return false;
+    }
+
+    const method = this.prototype[methodName];
+    if (typeof method !== 'function') {
+      return false;
+    }
+
+    const stub = SchemaVisitor.prototype[methodName];
+    if (method === stub) {
+      // If this.prototype[methodName] was just inherited from SchemaVisitor,
+      // then this class does not really implement the method.
+      return false;
+    }
+
+    return true;
+  }
+
+  // Concrete subclasses of SchemaVisitor should override one or more of these
+  // visitor methods, in order to express their interest in handling certain
+  // schema types/locations.
+  /* tslint:disable:no-empty */
+  public visitSchema(schema: GraphQLSchema) {}
+  public visitScalar(scalar: GraphQLScalarType) {}
+  public visitObject(object: GraphQLObjectType) {}
+  public visitFieldDefinition(field: GraphQLField<any, any>, details: {
+    objectType: GraphQLObjectType | GraphQLInterfaceType,
+  }) {}
+  public visitArgumentDefinition(argument: GraphQLArgument, details: {
+    field: GraphQLField<any, any>,
+    objectType: GraphQLObjectType | GraphQLInterfaceType,
+  }) {}
+  public visitInterface(iface: GraphQLInterfaceType) {}
+  public visitUnion(union: GraphQLUnionType) {}
+  public visitEnum(type: GraphQLEnumType) {}
+  public visitEnumValue(value: GraphQLEnumValue, details: {
+    enumType: GraphQLEnumType,
+  }) {}
+  public visitInputObject(object: GraphQLInputObjectType) {}
+  public visitInputFieldDefinition(field: GraphQLInputField, details: {
+    objectType: GraphQLInputObjectType,
+  }) {}
+  /* tslint:enable:no-empty */
+}
+
+// Generic function for visiting GraphQLSchema objects.
+export function visitSchema(
+  schema: GraphQLSchema,
+  // To accommodate as many different visitor patterns as possible, the
+  // visitSchema function does not simply accept a single instance of the
+  // SchemaVisitor class, but instead accepts a function that takes the
+  // current VisitableSchemaType object and the name of a visitor method and
+  // returns an array of SchemaVisitor instances that implement the visitor
+  // method and have an interest in handling the given VisitableSchemaType
+  // object. In the simplest case, this function can always return an array
+  // containing a single visitor object, without even looking at the type or
+  // methodName parameters. In other cases, this function might sometimes
+  // return an empty array to indicate there are no visitors that should be
+  // applied to the given VisitableSchemaType object. For an example of a
+  // visitor pattern that benefits from this abstraction, see the
+  // SchemaDirectiveVisitor class below.
+  visitorSelector: (
+    type: VisitableSchemaType,
+    methodName: string,
+  ) => SchemaVisitor[],
```

Comment:
After digesting the comment above, check out this very simple [`visitorSelector` implementation](https://github.com/apollographql/graphql-tools/blob/cb57840ba369ff15f6f7d6a720a3e63cd4dd2e04/src/test/testDirectives.ts#L306-L309) and then have a look at the more powerful [`visitorSelector`](https://github.com/apollographql/graphql-tools/blob/cb57840ba369ff15f6f7d6a720a3e63cd4dd2e04/src/schemaVisitor.ts#L339) used by `SchemaDirectiveVisitor.visitSchemaDirectives`.